fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -204,9 +207,10 @@ void FixedLagSmoother::optimizationLoop()
       // Optimization is complete. Notify all the things about the graph changes.
       notify(std::move(new_transaction), graph_->clone());
       // Compute a transaction that marginalizes out those variables.
+      lag_expiration = computeLagExpirationTime();
```

Comment:
Why is this done here, and not before `processQueue`?

It sems to me that the `lag_expiration` should be computed before, so it's updated after a **reset**:
https://github.com/locusrobotics/fuse/blob/3b59771f2a71e3e67c6cc5f063b7d7c6ef716ad0/fuse_optimizers/src/fixed_lag_smoother.cpp#L398

Maybe `lag_expiration` must be an attribute. :thinking:

Reply:
So.....

A more straight-forward processing regime would be:
* Compute lag expiration
* Compute marginal (This transaction will removes variables before the lag expiration)
* Add the marginal transaction to the graph (This actually removing those old variables)
* Add new transactions to the graph (As long as they don't involve old variables)
* Optimize the graph
* Notify publishers

However, that puts additional marginalization processing at the top of the loop. In an attempt to reduce publication latency, I shifted the above sequence by half a loop.
On loop 1:
* Add new transactions to the graph
* Optimize the graph
* Notify publishers
* Compute lag expiration
* Compute marginal (Note that we do not apply the marginal here)

Then on loop 2+:
* Add the marginal transaction to the graph (removing the old variables computed during the previous iteration)
* Add new transactions to the graph (as long as they don't involve those old variables)
* Optimize the graph
* Notify publishers
* Compute lag expiration
* Compute the next marginal

In the above scheme, the marginalization processing happens after the publishers are notified about the new graph values, under the theory that we have some spare time before the next optimization loop is triggered.

I could apply the marginal transaction at the end of the loop without optimizing the graph. That would clean up the code a bit, but it would leave the graph object in a different state than what was sent to the publishers, etc. I think that would be fine. But that does not change where in the loop we compute the lag expiration time.

The question you raise is, can we do this instead:
* Compute lag expiration
* Add new transactions to the graph
* Optimize the graph
* Notify publishers
* Compute marginal
* Add the marginal transaction to the graph

---

Existing sequence, iteration 1:
* Add t1, t2, t3 to the graph
* Optimize and publish (graph contains t1, t2, t3)
* expiration = t1
* Compute marginal for t1

Existing sequence, iteration 2:
* Apply marginal for t1
* Add new after t1 (t3, t4, t5)
* Optimize and publish  (graph contains t2, t3, t4, t5)
* expiration = t3
* Compute marginal for t2, t3

Existing sequence, iteration 3:
* Apply marginal for t2, t3
* Add new after t3 (t6, t7, t8)
* Optimize and publish (graph contains t4, t5, t6, t7, t8)
* expiration = t6
* Compute marginal for t4, t5, t6

---

Proposed sequence, iteration 1:
* expiration = t0 (The graph is empty)
* Add t1, t2, t3 to the graph
* Optimize and publish (graph contains t1, t2, t3)
* Compute marginal for t0 (which is empty)
* Apply marginal for t0 (which is empty)

Proposed sequence, iteration 2:
* expiration = t1 (Graph contains t1, t2, t3)
* Add new after t1 (t3, t4, t5)
* Optimize and publish (graph contains t1, t2, t3, t4, t5)
* Compute marginal for t1
* Apply marginal for t1 (Graph contains t2, t3, t4, t5)

Proposed sequence, iteration 3:
* expiration = t3 (Graph contains t2, t3, t4, t5)
* Add new after t3 (t6, t7, t8)
* Optimize and publish (graph contains t2, t3, t4, t5, t6, t7, t8)
* Compute marginal for t2, t3
* Apply marginal for t1, t2, t3 (Graph contains t4, t5, t6, t7, t8)

---

Assuming I did the above correctly, the proposed sequence ends each iteration with the same graph state, but results in more variables than needed during each optimization call. Unless there is a compelling reason to move that computation, I feel like the current sequence is correct.

But I agree there is bug relating to the reset function.

Reply:
> Assuming I did the above correctly, the proposed sequence ends each iteration with the same graph state, but results in more variables than needed during each optimization call. Unless there is a compelling reason to move that computation, I feel like the current sequence is correct.

Thanks for the detailed explanation. LGTM.

> But I agree there is bug relating to the reset function.

Actually, it looks like even if the `lag_expiration` is not reset, the logic should still work, since the transactions' timestamps would still be newer than it. At least, in my tests with #176 it looks like everything was already fine and no change was required in this regard.

Reply:
Something bad could probably still happen if you reset a bagfile or something and time moves backwards. I'll put in a fix regardless; it should be reset for consistency.

Reply:
Makes sense. :+1: